pkg/hooks/monitor.go

```diff
@@ -48,18 +49,47 @@ func (m *Manager) Monitor(ctx context.Context, sync chan<- error) {
 	for {
 		select {
 		case event := <-watcher.Events:
-			if event.Op&fsnotify.Remove == fsnotify.Remove {
-				ok := m.remove(filepath.Base(event.Name))
-				if ok {
-					logrus.Debugf("removed hook %s", event.Name)
+			filename := filepath.Base(event.Name)
+			if len(m.directories) <= 1 {
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					ok := m.remove(filename)
+					if ok {
+						logrus.Debugf("removed hook %s", event.Name)
+					}
+				} else if event.Op&fsnotify.Create == fsnotify.Create || event.Op&fsnotify.Write == fsnotify.Write {
+					err = m.add(event.Name)
+					if err == nil {
+						logrus.Debugf("added hook %s", event.Name)
+					} else if err != ErrNoJSONSuffix {
+						logrus.Errorf("failed to add hook %s: %v", event.Name, err)
+					}
 				}
-			}
-			if event.Op&fsnotify.Create == fsnotify.Create || event.Op&fsnotify.Write == fsnotify.Write {
-				err = m.add(event.Name)
-				if err == nil {
-					logrus.Debugf("added hook %s", event.Name)
-				} else if err != ErrNoJSONSuffix {
-					logrus.Errorf("failed to add hook %s: %v", event.Name, err)
+			} else if event.Op&fsnotify.Create == fsnotify.Create || event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Remove == fsnotify.Remove {
+				err = nil
+				found := false
+				for i := len(m.directories) - 1; i >= 0; i-- {
+					path := filepath.Join(m.directories[i], filename)
+					err = m.add(path)
+					if err == nil {
+						found = true
+						logrus.Debugf("(re)added hook %s (triggered activity on %s)", path, event.Name)
+						break
+					} else if err == ErrNoJSONSuffix {
+						found = true
+						break // this is not going to change for fallback directories
+					} else if os.IsNotExist(err) {
+						continue // move on to the next fallback directory
+					} else {
+						found = true
+						logrus.Errorf("failed to (re)add hook %s (triggered by activity on %s): %v", path, event.Name, err)
+						break
+					}
+				}
+				if (found || event.Op&fsnotify.Remove == fsnotify.Remove) && err != nil {
```

Comment:
Is this intended to be `!found`? Otherwise, it seems like we always remove hooks that we added above, immediately after they were added?

Reply:
> Is this intended to be `!found`? Otherwise, it seems like we always remove hooks that we added above, immediately after they were added?

The guard againg that is the trailing `&& err != nil`, because in the absence of a failed `add`, `err` will be `nil`.  The `found` clause here is avoiding a `remove` call if we didn't find the hook *and* it was a create or write event.  I don't see why we'd need a remove call for that case, but it shouldn't hurt to call it anyway.  Would you like me to transition to `if err != nil`?

And the unit tests prove we don't immediately remove hooks after adding them.  If the are more cases you'd like tested, let me know.

Reply:
Ahhh, didn't catch that. Alright, makes more sense that way.